### PR TITLE
`Iris`: Allow instructors to adapt programming exercises with the help of GPT-4

### PR DESCRIPTION
**Pull request by MichaelOwenDyer**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I added pre-authorization annotations according to the [guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/#rest-endpoint-best-practices-for-authorization) and checked the course groups for all new REST Calls (security).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client-tests/).
- [x] I added `authorities` to all new routes and checked the course groups for displaying navigation elements (links, buttons).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

As part of our thesis project, @RY997 and I are adding the Iris AI chatbot to the Code Editor screen to offer assistance with creating and adapting programming exercises.

### Steps for Testing

Prerequisites:
- 1 Editor
- 1 Programming Exercise

1. Log in to Artemis **on TS9**
2. Navigate to Server Administration -> Iris Settings
3. Scroll halfway down to "Code Editor Settings", check that the feature is enabled and that there is at least one AI model checked
4. Navigate to Course Management and go into the course you would like to test the feature in
5. Open the course Iris settings, enable the Iris Code editor feature there as well
6. Open a programming exercise, you should see the Iris logo appear in the top right
7. Click on the Iris logo to open the chat window, have a conversation with Iris about the exercise
8. Ask Iris to change the exercise somehow
9. Iris should respond with a plan, check that the plan content matches your requirements
10. Click the "Execute" button to have Iris start working on the plan step by step
11. Check that the step colors are updated to reflect their progress
12. Check that when a step turns green, the changes it describes have actually been applied to the corresponding exercise component

#### Performance Review
- [x] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

Client tests
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| http-code-editor-session.service.ts | 100% | ✅               |
| http-chat-session.service.ts | 85% | ✅               |
| http-session.service.ts | 100% | ✅               |
| http-code-editor-message.service.ts | 100% | ✅               |
| http-chat-message.service.ts | 100% | ✅               |
| http-message.service.ts | 92% | ✅               |
| code-editor-session.service.ts | 96% | ✅               |
| chat-websocket.service.ts | 94% | ✅               |
| websocket.service.ts | 87% | ✅               |
| code-editor-chatbot-button.component.ts | 100% | ✅               |
| tutor-chatbot-button.component.ts | 100% | ✅               |
| chatbot-button.component.ts | 95% | ✅               |
| code-editor-instructor-and-editor-container.component.ts | 93% | ✅              |
| chatbot-widget.component.ts | 73%(covers all newly created methods) | ✅              |

Server tests
| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| IrisCodeEditorSessionResource.java | 100% | ✅               |
| IrisChatSessionResource.java | 100% | ✅               |
| IrisCodeEditorMessageResource.java | 87% | ✅               |
| IrisCodeEditorSessionService.java | 81% | ✅               |

### Screenshots

Screenshot 2023-11-24 at 17 56 02
Screenshot 2023-11-24 at 17 57 03

### CODE REVIEW

**Review by Jan-Thurner — APPROVED** [review]
Tested in a Testing session on TS9. 

A small bug was found: Error messages don't disappear when retrying a failed task. 
Not all LoC could be reviewed due to the shier size of the PR.

**Review by JanaNF — DISMISSED** [review]
Tested in a Testing session on TS9.

**Review by sarpsahinalp — DISMISSED** [review]
Tested in a Testing session on TS9.

**Review by vinceclifford — DISMISSED** [review]
Tested during testing session

**Review by florian-glombik — APPROVED** [review]
I did only code review 12/120 files that seemed to be related to other stuff. I think it would be good to get this initial state merged to handle upgrades in smaller steps.

There are some issues that can be fixed in a follow up and smaller steps: 
* error message that does not disappear when clearing the chat
* error message `no LLM available` - I think this can be refined to also tell the user to not try again right now (was related to a rate limit in our case)

File: src/main/java/de/tum/in/www1/artemis/service/iris/IrisConstants.java [hunk 1]
```diff
@@ -7,106 +7,136 @@ public final class IrisConstants {
...
+                {{gen 'revision' temperature=0.2 max_tokens=2000}}
             {{~/assistant}}
-                    """;
+
+            {{#if (equal revision " ")}}
+                {{set 'response' responsedraft}}
+            {{else}}
+                {{set 'response' revision}}
+            {{/if}}
+            """;
```
Comment (Hialus) [inline src/main/java/de/tum/in/www1/artemis/service/iris/IrisConstants.java:139]: Please revert this change. It breaks the chat in certain cases and does not really offer much of an improvement when it works.

**Review by Hialus — CHANGES_REQUESTED** [review]

**Review by Hialus — DISMISSED** [review]
Thanks for fixing it 👍🏻

**Review by Hialus — APPROVED** [review]
